proxy/logging/LogObject.cc

```diff
@@ -477,6 +481,89 @@ LogObject::log(LogAccess *lad, const char *text_entry)
   return this->log(lad, std::string_view{text_entry ? text_entry : ""});
 }
 
+class ThreadLocalLogBufferManager : public Continuation
+{
+public:
+  ThreadLocalLogBufferManager()
+  {
+    this->thread_affinity = this_ethread();
+    SET_HANDLER(&ThreadLocalLogBufferManager::wakeup);
+
+    int period = Log::config->max_secs_per_buffer >> 1;
+    if (period < 1) {
+      period = 1;
+    }
+    eventProcessor.schedule_every(this, period * HRTIME_SECOND, ET_CALL);
+    Debug("log-config", "thread local buffer manager init: %d wakeup period", period);
+  }
+  ~ThreadLocalLogBufferManager() override
+  {
+    Debug("log-config", "thread local buffer manager destructor");
+    // only the LogBuffer objects are owned by this
+    for (auto [o, b] : current_buffers) {
+      // ideally we flush these here but there are shutdown order issues so if the
+      // logbuffer still exists at this point we have to drop it
+      delete b;
+    }
+    current_buffers.clear();
+  }
+  int
+  wakeup(int /* event ATS_UNUSED */, Event * /* e ATS_UNUSED */)
+  {
+    for (auto &[o, b] : current_buffers) {
+      if (b && ink_hrtime_to_sec(Thread::get_hrtime()) > b->expiration_time()) {
+        o->flush_buffer(b);
+        b = nullptr;
+      }
+    }
+
+    return EVENT_CONT;
+  }
+
+  LogBuffer *
+  current_buffer(LogObject *o, size_t *offset, size_t bytes_needed)
+  {
+    LogBuffer *buffer = current_buffers[o];
+    if (buffer == nullptr) {
+      buffer             = new LogBuffer(Log::config, o, Log::config->log_buffer_size);
+      current_buffers[o] = buffer;
+    }
+    if (buffer->add_entry(offset, bytes_needed) != LogBuffer::LB_OK) {
+      o->flush_buffer(buffer);
+
+      buffer             = new LogBuffer(Log::config, o, Log::config->log_buffer_size);
+      current_buffers[o] = buffer;
+      if (buffer->add_entry(offset, bytes_needed) != LogBuffer::LB_OK) {
+        return nullptr;
+      }
+    };
+    return buffer;
+  }
+
+private:
+  std::map<LogObject *, LogBuffer *> current_buffers;
```

Comment:
Seems like this could be an unordered map.

Reply:
Why is it better as an ordered map?

Reply:
Since logging is per-transaction, seems like it's worth making the most optimal choice

Reply:
My other comments are about judgement calls, and I don't use reviews to second guess, but I think it's worth knowing your reasons here.

Reply:
LogObjects are created at startup per log, and then more won't be created.  Therefore, this map will just grow to a small fixed size and then only the buffer pointer will change as log buffers are filled up.  My presumption is the overhead of finding the entry will be negligible with either tree or table storage. I briefly considered a vector to store them since the lookup would be practically constant time.

Reply:
If you prefer an unordered_map, I don't mind changing it to that.

Reply:
Sounds reasonable, I'll approve as soon as the docs problem is resolved.